processing/src/main/java/io/druid/segment/data/CompressedVSizeIntsIndexedSupplier.java

```diff
@@ -0,0 +1,423 @@
+import java.util.List;
+
+public class CompressedVSizeIntsIndexedSupplier implements WritableSupplier<IndexedInts>
+{
+  public static final byte version = 0x2;
+
+  private final int totalSize;
+  private final int sizePer;
+  private final int numBytes;
+  private final int bigEndianShift;
+  private final int littleEndianMask;
+  private final GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers;
+  private final CompressedObjectStrategy.CompressionStrategy compression;
+
+  CompressedVSizeIntsIndexedSupplier(
+      int totalSize,
+      int sizePer,
+      int numBytes,
+      GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers,
+      CompressedObjectStrategy.CompressionStrategy compression
+  )
+  {
+    this.totalSize = totalSize;
+    this.sizePer = sizePer;
+    this.baseBuffers = baseBuffers;
+    this.compression = compression;
+    this.numBytes = numBytes;
+    this.bigEndianShift = Integer.SIZE - (numBytes << 3); // numBytes * 8
+    this.littleEndianMask = (int)((1L << (numBytes << 3)) - 1); // set numBytes * 8 lower bits to 1
+  }
+
+  public static int maxIntsInBufferForBytes(int numBytes)
+  {
+    return (CompressedPools.BUFFER_SIZE - bufferPadding(numBytes)) / numBytes;
+  }
+
+  private static int bufferPadding(int numBytes)
+  {
+    // when numBytes == 3 we need to pad the buffer to allow reading an extra byte
+    // beyond the end of the last value, since we use buffer.getInt() to read values.
+    // for numBytes < 3 we remove the need for padding by reading short or bytes directly.
+    if (numBytes == 3) {
+      return Ints.BYTES - numBytes;
+    } else {
+      return 0;
+    }
+  }
+
+  public static int maxIntsInBufferForValue(int maxValue)
+  {
+    return maxIntsInBufferForBytes(VSizeIndexedInts.getNumBytesForMax(maxValue));
+  }
+
+  public int size()
+  {
+    return totalSize;
+  }
+
+  @Override
+  public IndexedInts get()
+  {
+    final boolean powerOf2 = sizePer == (1 << Integer.numberOfTrailingZeros(sizePer));
+
+    if(powerOf2) {
+      if(numBytes == Ints.BYTES) {
+        return new CompressedFullSizeIndexedInts() {
+          @Override
+          public int get(int index)
+          {
+            return get2(index);
+          }
+        };
+      } else if(numBytes == Shorts.BYTES) {
+        return new CompressedShortSizeIndexedInts() {
+          @Override
+          public int get(int index)
+          {
+            return get2(index);
+          }
+        };
+      } else if (numBytes == 1) {
+        return new CompressedByteSizeIndexedInts()
+        {
+          @Override
+          public int get(int index)
+          {
+            return get2(index);
+          }
+        };
+      } else {
+        return new CompressedVSizeIndexedInts()
+        {
+          @Override
+          public int get(int index)
+          {
+            return get2(index);
+          }
+        };
+      }
+    } else {
+      // optimized versions for 1, 2, and 4 byte indices
```

Comment:
Aren't 1, 2 and 4 the same as the 1, Shorts.BYTES and Ints.BYTES from the part above?

This code is really confusing me...